---
title: "Xenium 0724"
output: html_document
date: "2024-07-22"
---

```{r}
library(Seurat)
library(tidyverse)
library(future)


```

```{r}

# Loading the data from the folder generated by Xenium onboard analysis. I named the path of the folder containing all the information from the run. You could specify assay= xenium when you LoadXenium but I did not do that here so ut loaded all the assays available from the data which include data from the control probes etc. 

path <- "C:/Users/nanah/OneDrive - Northwestern University/Laptop files/OneDrive - Northwestern University/Berdnikovs lab/output-XETG00191__0018632__Berdnikovs_mLung_2__20240703__184037"


# Load the Xenium data
xenium.obj <- LoadXenium(path, fov = "fov")

# Viewing all the genes captured. There are 379 genes represented in this sample.

row.names(xenium.obj)

# remove cells with 0 counts
xenium.obj <- subset(xenium.obj, subset = nCount_Xenium > 0)

```

```{r}

# Here are violin plots of genes per cell (nFeature_Xenium) and transcript counts per cell (nCount_Xenium)
VlnPlot(xenium.obj, features = c("nFeature_Xenium", "nCount_Xenium"), ncol = 2, pt.size = 0)

# Next we can look at the position of some markers from the gene panel using ImageDimplot()

ImageDimPlot(xenium.obj, fov = "fov", molecules = c("Vwf", "Mmrn1", "Scgb3a2", "Cd3d"), nmols = 20000, dark.background   = FALSE)

# We can also visualize the expression of genes at the per-cell level using a variation of FeaturePlot(). The max.cutoff was adjusted to improve contrast of the feature plots eg. max.cutoff = 'q90'. This is somthing that can be tweaked based on the gene you are trying to visualize and the sample. 

ImageFeaturePlot(xenium.obj, features = c("Epcam" , "Tagln", "Alox5ap", "Thbs4" , "Mpo" ,"Csf3r"), max.cutoff = c(25,
    35, 12, 10), size = 0.75, cols = c("white", "red"))

ImageFeaturePlot(xenium.obj, features = "nCount_Xenium") + theme(legend.position = "right")

```


```{r}
# Next we can use SCTrasnform for normalization followed by standard dimension reduction and clustering. This step takes a few minutes. Then we use DimPlot() to visualize the results of the clustering by coloring each cell according to its cluster in UMAP space. You can also use ImageDimPlot() to overlay the clustering on the image. 

xenium.obj <- SCTransform(xenium.obj, assay = "Xenium")
xenium.obj <- RunPCA(xenium.obj, npcs = 48, features = rownames(xenium.obj))
xenium.obj <- RunUMAP(xenium.obj, dims = 1:40)
xenium.obj <- FindNeighbors(xenium.obj, reduction = "pca", dims = 1:40)
xenium.obj <- FindClusters(xenium.obj, resolution = 0.75)


DimPlot(xenium.obj, label = TRUE # , raster = FALSE
        )

ImageDimPlot(xenium.obj)

# or

ImageDimPlot(xenium.obj, cols = "polychrome", size = 0.75)

colorz<- c("black", "green4", "#442288","yellow", "#6CA2EA", "#B5D33D", "magenta",  "#EB7D5B", "red2",  "orange", "hotpink", "gold", "darkorange", "saddlebrown", "gray35", "blue4", "springgreen3", "turquoise4", "greenyellow", "darkred", "cadetblue1", "darkslategrey", "plum4", "darkmagenta", "khaki2", "rosybrown1", "palegreen3", "slateblue4", "hotpink3" )

DimPlot(new_xenium.obj, cols = colorz, size = 1, group.by = "celltype", raster = FALSE)


```


```{r}
# After normalization and generating the clusters, we can visulize the expression of the markers we previously looked at and assign cluster labels as you see fit. 

ImageFeaturePlot(xenium.obj, features = c("Cxcr2", "S100a8", "Ccr3", "Siglecf", "Alox15", "Cd14", "Mylk"))



VlnPlot(xenium.obj, features = c("Cd8a", "Ccr7", "Laptm5", "Rac2", "Plbd1", "Itgb7", "Ms4a6c"), raster = FALSE)


```


```{r}
## run PrepSCTFindMarkers on merged object

Deg <- FindAllMarkers(xenium.obj, test.use = "roc", assay = "SCT")

```

```{r}
##Assigning cell identity to clusters 

new.cluster.id <- c("gCap", "B-Cells", "AT2", "Mesenchyme", "CD8 T-cells", "Monocyte-derived Macs", "Cd101 Neutrophils", "Myeloid precursors", "Plasma cells", "Eosinophils", "Myofibroblasts", "Neurons", "Smooth muscle, Alveolar", "Angiogenesis", "AT1", "Cilliated epithelium", "aCap", "cDCs", "Endothelium", "Esophagus lining", "Lung Pleural Lining", "Tregs", "CD4 T-cells", "Esophagus (muscle)", "Neutrophil 2", "pDCs", "Platelets", "Adipose", "M2 macrophages")

new_xenium.obj <- xenium.obj

names(new.cluster.id) <- levels(new_xenium.obj)

new_xenium.obj<- RenameIdents(new_xenium.obj, new.cluster.id)

new_xenium.obj$celltype <- Idents(new_xenium.obj)


DimPlot(new_xenium.obj, label = TRUE, reduction = "umap" # , raster = FALSE
        )

```


```{r}
DefaultAssay(new_xenium.obj) <- 'Xenium'

ImageFeaturePlot(new_xenium.obj,  fov = "fov", features =  "Alox15", axes = TRUE, mols.size = 5, nmols = 20000, size = 1.5, max.cutoff = "q90")

```

```{r}
VlnPlot(xenium.obj, features = c("Cd79b", "Csf3r", "Siglecf", "Alox15", "Oasl1", "Rsad2"), raster = FALSE)
```

```{r}

crop <- Crop(xenium.obj[["fov"]], x = c(2500, 3500 ), y = c(3600, 4600))
xenium.obj[["crop"]] <- crop
p2 <- ImageFeaturePlot(xenium.obj, fov = "crop", features = "Oasl1", size = 1, axes = TRUE, max.cutoff = 1)
p2

```

```{r}

new_xenium.obj <- BuildNicheAssay(object = new_xenium.obj, fov = "fov", group.by = "celltype",
    niches.k = 10, neighbors.k = 30)


celltype.plot <- ImageDimPlot(new_xenium.obj, group.by = "celltype", size = 1.75, #cols = "polychrome",
    dark.background = F) + ggtitle("Cell type") +
    scale_fill_manual(values = c("black", "green4", "#442288","yellow", "#6CA2EA", "#B5D33D", "magenta",  "#EB7D5B", "red2",  "orange", "hotpink", "gold", "darkorange", "saddlebrown", "gray35", "blue4", "springgreen3", "turquoise4", "greenyellow", "darkred", "cadetblue1", "darkslategrey", "plum4", "darkmagenta", "khaki2", "rosybrown1", "palegreen3", "slateblue4", "hotpink3" ))
niche.plot <- ImageDimPlot(new_xenium.obj, group.by = "niches", size = 1.75, dark.background = F) + ggtitle("Niches") +
    scale_fill_manual(values = c("#442288", "green4", "red", "#6CA2EA","black" , "gold4" , "magenta", "blue", "darkorange4" , "yellow2"  ))
celltype.plot | niche.plot

```

```{r}

Niche.table <- table (new_xenium.obj$celltype, new_xenium.obj$niches)

Niche.table
```

```{r}

## Niche diff analysis, set ident(...) to "niche" and run find all markers
 Idents(new_xenium.obj) <- "niches"
Deg_niche <- FindAllMarkers(new_xenium.obj, assay = "SCT")


Sig.niche<-Deg_niche %>%
  group_by(cluster) %>%
  filter(p_val_adj<0.01, avg_log2FC>2, pct.1>0.2 )
DoHeatmap(object = new_xenium.obj, features = Sig.niche$gene, assay = "SCT") + scale_fill_gradient2(low = "black", mid = "darkblue", high= "orangered", midpoint = 0)

DoHeatmap(object = new_xenium.obj, features = Marker_gene, assay = "SCT") + scale_fill_gradient2(low = "black", mid = "darkblue", high= "orangered", midpoint = 0)


```
```{r}
## Let's get the scaled/log-transformed counts of these genes from the Seurat object
## use fetch data when you had to join layers in seurat v5
Marker_gene<- c("Cxcr2", "S100a8", "Slc7a11", "Mmp9", "Csf3r", "Nlrp3", "Siglece", "Cd200r1", "Cd300lf", "Cstdc4", "Itgax", "Il1r2", "Mmp8", "Cd101", "Ctsd", "Lyz2", "Ccl2", "Kit", "Sftpd", "Car4", "Tnc", "Ager", "Runx2", "Rsad2", "Sftpb", "Pdgfb", "Sfta2", "Kitl")

DefaultAssay(new_xenium.obj) <- 'Xenium'

exp_mat <- FetchData(object = new_xenium.obj, vars = Marker_gene)

## Let's get cell metadata now

meta <-new_xenium.obj@meta.data %>% 
  select(niches)

## Merge in the expression data. Just need to transpose it so that cells are down the rows.
## Use this after running fetch data. If you did not run fetch data, exp_mat would need to be in a data frame.

meta <- bind_cols(meta, exp_mat)
head(meta) #View the first few lies

## To work with ggplot2, we need to get our data in long/tidy format

meta <- pivot_longer(meta, -niches, names_to="Gene", values_to="Expression")

meta$niches <- as.character(meta$niches)

meta_summary <- meta %>%
  group_by(niches, Gene) %>%
  summarise(Avg = mean(Expression),
            Pct = sum(Expression > 0) / length(Expression) * 100)


## Plot function but first keep the gene list order

meta_summary$Gene <- factor(meta_summary$Gene, levels=unique(Marker_gene))


heatmap <- ggplot(meta_summary, aes(x=Gene, y=niches, fill = Avg ))+  geom_tile(color = "black") +
  scale_fill_gradient2(low = "black", mid = "blue", high= "gold", midpoint = 2)+
  #scale_fill_gradientn( colors = viridis(n = 4,  option = "cividis" )) +
  coord_fixed()+
 theme(axis.text.x = element_text(size=14, angle=45, hjust=1, color="black"),
        axis.text.y = element_text(size=15, color="black"),
        axis.title = element_text(size=14))
heatmap


dot_plot <- ggplot(meta_summary, aes(x=Gene, y=niches)) +
  geom_point(aes(size = Avg, fill = Pct), color="black", shape=21) +
  scale_size("% detected", range = c(0,5)) +
  scale_fill_gradientn(colours = viridisLite::plasma(5),
                       guide = guide_colorbar(ticks.colour = "black",
                                              frame.colour = "black"),
                       name = "Average\nexpression") +
  ylab("Cluster") + xlab("") +
  theme_bw() +
  theme(axis.text.x = element_text(size=14, angle=45, hjust=1, color="black"),
        axis.text.y = element_text(size=15, color="black"),
        axis.title = element_text(size=14))

dot_plot

```


```{r}
VlnPlot(new_xenium.obj, features = c("Cs", "Oasl1", "Rsad2", "Ifit1", "Ifitm1", "Mmp9", "Fn1", "Tnc", "Dcn"), raster = FALSE)

```

```{r}
## Subseting the eosinophils in different niches. Starting with niche 3. I will first subset niche 3. Make sure Idents of the object is "niche". Then change te ident of the subsetted object to "cell type" and subset eos out of each niche. I am focusing on niche 3,6, and 10  
Idents(new_xenium.obj) <- "celltype"
Eos<- subset(new_xenium.obj, idents = 'Eosinophils' ) 

 Idents(Eos) <- "niches" 
  
niche3 <- subset(Eos, idents = '3' )
niche6 <- subset(Eos, idents = '6' )
niche10 <- subset(Eos, idents = '10' )


## Now we merge all 3 and run differential expression analysis

MergedEos <- merge(niche3, y= c(niche6, niche10))


MergedEos<- SCTransform(MergedEos, assay = "Xenium")
MergedEos<- RunPCA(MergedEos, npcs = 30, verbose = TRUE)
MergedEos<- RunUMAP(MergedEos, reduction = "pca", dims = 1:10)
MergedEos<- FindNeighbors(MergedEos, reduction = "pca", dims = 1:10)
MergedEos<- FindClusters(MergedEos, resolution = 0.1)
DimPlot(MergedEos, reduction = "umap", label = TRUE, repel = TRUE, group.by = "niches")

DimPlot(MergedEos, reduction = "umap", label = TRUE, repel = TRUE, cols = Merged)

DimPlot(MergedEos, reduction = "umap", label = FALSE, repel = TRUE, cols = c("red", "#6CA2EA", "#442288"), pt = 0.5, group.by = "niches")

Merged <- c("lightblue4", "brown2", "purple", "green2")

ImageDimPlot(MergedEos)

VlnPlot(MergedEos, features = "Alox5ap")

MergedEos<- JoinLayers(MergedEos, assay = "Xenium")

## For some reason, SCTransfrom changes the "umi.assay" slot of all the SCTAssay objects in a merged object to RNA so you have yo change it back one by one. I only have 3 objects in this merged object so I change 1,2 and 3. 
slot(object = MergedEos@assays$SCT@SCTModel.list[[1]], name = "umi.assay") <- "Xenium"
slot(object = MergedEos@assays$SCT@SCTModel.list[[2]], name = "umi.assay") <- "Xenium"
slot(object = MergedEos@assays$SCT@SCTModel.list[[3]], name = "umi.assay") <- "Xenium"

MergedEos <- PrepSCTFindMarkers(MergedEos)
roc.Eos<- FindAllMarkers(MergedEos, assay = "SCT" )

Idents(MergedEos) <- "seurat_clusters" 
Niche.Eos<- FindAllMarkers(MergedEos, assay = "SCT" )
 
```

```{r}
library(dplyr)
Sig.Eos<-roc.Eos %>%
  group_by(cluster) %>%
  filter(p_val_adj<0.01, avg_log2FC>1.1, pct.1>0.12 )
DoHeatmap(object = MergedEos, features = Sig.Eos$gene, assay = "SCT", group.colors = Merged) + scale_fill_gradient2(low = "black", mid = "darkblue", high= "orangered", midpoint = 0)



## Let's get the scaled/log-transformed counts of these genes from the Seurat object
## use fetch data when you had to join layers in seurat v5

exp_mat <- FetchData(object = MergedEos, vars = Sig.Eos$gene)

## Let's get cell metadata now

meta <-MergedEos@meta.data %>% 
  select(seurat_clusters)

## Merge in the expression data. Just need to transpose it so that cells are down the rows.
## Use this after running fetch data. If you did not run fetch data, exp_mat would need to be in a data frame.

meta <- bind_cols(meta, exp_mat)
head(meta) #View the first few lies

## To work with ggplot2, we need to get our data in long/tidy format

meta <- pivot_longer(meta, -seurat_clusters, names_to="Gene", values_to="Expression")

meta_summary <- meta %>%
  group_by(seurat_clusters, Gene) %>%
  summarise(Avg = mean(Expression),
            Pct = sum(Expression > 0) / length(Expression) * 100)


## Plot function but first keep the gene list order

meta_summary$Gene <- factor(meta_summary$Gene, levels=unique(genes))


heatmap <- ggplot(meta_summary, aes(x=Gene, y=seurat_clusters, fill = Avg ))+  geom_tile(color = "black") +
  scale_fill_gradient2(low = "black", mid = "blue", high= "gold", midpoint = 2)+
  #scale_fill_gradientn( colors = viridis(n = 4,  option = "cividis" )) +
  coord_fixed()+
 theme(axis.text.x = element_text(size=14, angle=45, hjust=1, color="black"),
        axis.text.y = element_text(size=15, color="black"),
        axis.title = element_text(size=14))
heatmap

```

```{r}

# Proportion / cell number composition per cluster
ggData = data.frame(prop.table(table(MergedEos$niches, MergedEos$seurat_clusters), margin = 2))
colnames(ggData) = c("library", "cluster", "value")
p1 <- ggplot(ggData, aes(cluster, value, fill = library)) +
  geom_col() + xlab("Cluster") + ylab("Proportion of Cells (%)") +   scale_fill_manual(values = c("red", "#6CA2EA", "#442288")) + coord_flip()+
  theme(
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')
       )
p1



library(pheatmap)
library(RColorBrewer)
```

```{r}

col.niche<- c(  "blue", "darkorange4","red","black" ,"gold4" ,"#6CA2EA", "magenta",  "green4", "yellow2" , "#442288")

ggData2 = data.frame(prop.table(table(new_xenium.obj$niches, new_xenium.obj$celltype), margin = 2))
colnames(ggData2) = c("library", "cluster", "value")
p1 <- ggplot(ggData2, aes(cluster, value, fill = library)) +
  geom_col() + xlab("Cluster") + ylab("Proportion of Cells (%)") +  scale_fill_manual(values = col.niche) + coord_flip()+
  theme(
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')
       )
p1



```

```{r}
categories <- c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")
Eos.prop <- c(2.29, 0.17, 17.03, 4.18, 4.52, 32.07, 7.49, 7.05, 0.17, 25.03)

Neu1.prop <- c(0.88,	0.00,	22.98,	1.30,	3.55,	6.41,	3.56,	3.86,	0.00,	57.46
)

Neu2.prop <- c(2.50,	0.06,	45.09,	0.45,	9.38,	5.39,	2.89,	1.93,	0.06,	32.24)

MP.prop <- c(0.38,	0.01,	30.33,	0.67,	6.09,	4.76,	3.16,	1.14,	0.05,	53.41
)

df <- data.frame(Category = categories, Proportion = Eos.prop)

col.pie<- c(  "blue", "#442288", "darkorange4","red","black" ,"gold4" ,"#6CA2EA", "magenta",  "green4", "yellow2"  )

## DOnut plot
  ggplot(df, aes(x = 3, y = Proportion, fill = Category)) + 
  geom_bar(stat = "identity", width = 1) + 
  coord_polar(theta = "y") + 
  scale_fill_manual(values = col.pie) + 
  xlim(0.5,4) +  # Create space for the hole
  theme_void() +  # Minimalist theme
  theme(legend.position = "right") +  # Adjust legend position
  labs(title = "Eos per niche")
  

  +
   geom_text(aes(label = paste0(round(Proportion, 1), "%")), 
            position = position_stack(vjust = 0.5), check_overlap = TRUE) 
  
  
#Pie chart
ggplot(df, aes(x = "", y = Proportion, fill = Category)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  scale_fill_manual(values = custom_colors) +
  geom_text(aes(label = label), position = position_stack(vjust = 0.5)) +  # Add labels
  theme_minimal()




```


```{r}
## Subseting the CD101 neutrophils in different niches. 

Idents(new_xenium.obj) <- "celltype"
Neu101<- subset(new_xenium.obj, idents = 'Cd101 Neutrophils' ) 

 Idents(Neu101) <- "niches" 
  
niche3 <- subset(Neu101, idents = '3' )
niche10 <- subset(Neu101, idents = '10' )


## Now we merge all 3 and run differential expression analysis

MergedNeu101 <- merge(niche3, y= c(niche10))


MergedNeu101<- SCTransform(MergedNeu101, assay = "Xenium")
MergedNeu101<- RunPCA(MergedNeu101, npcs = 30, verbose = TRUE)
MergedNeu101<- RunUMAP(MergedNeu101, reduction = "pca", dims = 1:10)
MergedNeu101<- FindNeighbors(MergedNeu101, reduction = "pca", dims = 1:10)
MergedNeu101<- FindClusters(MergedNeu101, resolution = 0.1)
DimPlot(MergedNeu101, reduction = "umap", label = TRUE, repel = TRUE, group.by = "niches")

DimPlot(MergedNeu101, reduction = "umap", label = TRUE, repel = TRUE, cols = Merged)

DimPlot(MergedNeu101, reduction = "umap", label = FALSE, repel = TRUE, cols = c("red", "#6CA2EA", "#442288"), pt = 0.5, group.by = "niches")

Merged <- c("lightblue4", "brown2", "purple", "green2")

ImageDimPlot(MergedNeu101)

VlnPlot(MergedNeu101, features = "Alox5ap")

MergedNeu101<- JoinLayers(MergedNeu101, assay = "Xenium")

## For some reason, SCTransfrom changes the "umi.assay" slot of all the SCTAssay objects in a merged object to RNA so you have yo change it back one by one. I only have 3 objects in this merged object so I change 1,2 and 3. 
slot(object = MergedNeu101@assays$SCT@SCTModel.list[[1]], name = "umi.assay") <- "Xenium"
slot(object = MergedNeu101@assays$SCT@SCTModel.list[[2]], name = "umi.assay") <- "Xenium"


MergedNeu101 <- PrepSCTFindMarkers(MergedNeu101)
roc.Neu101<- FindAllMarkers(MergedNeu101, assay = "SCT" )

Idents(MergedNeu101) <- "seurat_clusters" 
Niche.Neu101<- FindAllMarkers(MergedNeu101, assay = "SCT" )
 
```